### PR TITLE
[Reviewer: Andy] Optional Ralf

### PR DESCRIPTION
**Pull request by rkd-msw**
This fixes https://github.com/Metaswitch/chef/issues/151 (or completes the fix to https://github.com/Metaswitch/chef/issues/124, depending on how you look at it), by making Ralf nodes genuinely optional - you can remove all Ralf nodes from a deployment, or add them in to a Ralfless deployment, and /etc/clearwater/config's ralf_hostname will adjust accordingly.

It has two parts:
* code in 'knife box create' to include a ralf_hostname if it's a Ralf node or there's a --with-ralf option - which should be uncontroversial
* code in 'knife deployment resize' to spot if we've gone from having no Ralfs to having some, or vice-versa, and update /etc/clearwater/config and restart the process on Sprout/Bono/Ralf nodes to account for that.

This second part calls 'sudo monit restart all', which means this is service-affecting - I think that's fine, because it only happens when you add Ralfs to a deployment with none (or vice versa) which will be very rare. (It should be possible to make that non-service-affecting by doing a rolling quiesce/restart, but that will be slow, and I don't think it's necessary.)

I've
* created new nodes with 'knife box create sprout --with-ralf' and 'knife box create ralf' and verified that both had a ralf_hostname set
* spun up a deployment specifying '--ralf_count 1' and verified that Sprout and Ralf had a ralf_hostname
* removed the Ralf with 'knife deployment resize --ralf-count 0 --force' and verified that Sprout's config no longer included a Ralf hostname, and was restarted (so Ralf was not in the command-line args)
* spun up a deployment without specifying ralf_count and verified that Sprout had no ralf_hostname
* subsequently run 'knife deployment resize --ralf-count 1' and verified that Sprout gained a Ralf hostname, and was restarted (so Ralf was in the command-line args), and that Ralf came up correctly
* subsequently run 'knife deployment resize --ralf-count 0' and 'knife deployment resize --finish' and verified that Sprout's config no longer included a Ralf hostname, and was restarted (so Ralf was not in the command-line args)

It's possible to fix up existing deployments like dogfood just by running "knife deployment resize" with no changes - this triggers a check that the number of Ralf nodes is consistent with the ralf attribute on existing nodes, and sets the ralf attribute if it's needed.

### CODE REVIEW

**Comment by rkd-msw** [comment]
We discussed yesterday, and I think the outcome was that:

* having the actual Ralf hostname as the attribute would be more flexible, but isn't a useful change right now
* `update_ralf_hostname` could be made more generic (to accommodate other optional nodes). I'd like to leave as-is until we have another node type that needs `update_*_hostname`, and make it generic at that point.

File: cookbooks/clearwater/recipes/infrastructure.rb [hunk 1]
```diff
@@ -117,7 +117,7 @@
     cdf = "cdf." + domain
   end
 
-  ralf = if node[:clearwater][:ralf] and node[:clearwater][:ralf] > 0
+  ralf = if node[:clearwater][:ralf] and ((node[:clearwater][:ralf] == true) || (node[:clearwater][:ralf] > 0))
```
Comment (bossmc) [inline cookbooks/clearwater/recipes/infrastructure.rb]: Why do we pass the count of nodes down?  Why isn't this just a boolean?

Comment (bossmc) [inline cookbooks/clearwater/recipes/infrastructure.rb]: If it were we could simply use `if node[:clearwater][:ralf]`

File: plugins/knife/knife-deployment-resize.rb [hunk 2]
```diff
@@ -287,6 +289,36 @@ def get_current_counts
...
     end
 
+    def update_ralf_hostname environment, cloud
+      ralfs = find_nodes(roles: "clearwater-infrastructure", role: "ralf").length
+
+      changed_nodes = []
+
+      %w{bono ibcf sprout ralf}.each do |node_type|
+        find_nodes(roles: "clearwater-infrastructure", role: node_type).each do |node|
+          has_ralf = node.set[:clearwater][:ralf]
```
Comment (bossmc) [inline plugins/knife/knife-deployment-resize.rb]: You don't need (want) the `set` here.  That's only used for changing the value.

File: plugins/knife/knife-deployment-resize.rb [hunk 3]
```diff
@@ -287,6 +289,36 @@ def get_current_counts
...
+          puts "#{node.name}: ralf attribute is #{has_ralf} and number of ralfs is #{ralfs}"
+          if (ralfs == 0) && has_ralf
+            node.set[:clearwater][:ralf] = false
```
Comment (bossmc) [inline plugins/knife/knife-deployment-resize.rb]: Is `set` the correct level?  Should it be `default`?

File: plugins/knife/knife-deployment-resize.rb [hunk 4]
```diff
@@ -383,6 +415,7 @@ def run
           end
         end
 
+        update_ralf_hostname config[:environment], config[:cloud].to_sym
```
Comment (bossmc) [inline plugins/knife/knife-deployment-resize.rb]: Brackets (I know there are counter examples elsewhere in the file but that's not an excuse to make it worse :)).

**Comment by bossmc** [comment]
Reviewed (with the above caveats).

File: cookbooks/clearwater/recipes/infrastructure.rb (hunk 1 above)
Comment (rkd-msw) [inline cookbooks/clearwater/recipes/infrastructure.rb]: This pull request changes the `:ralf` attribute from a count to a boolean, so we need to handle both cases.

File: plugins/knife/knife-deployment-resize.rb [hunk 5]
```diff
@@ -287,6 +289,36 @@ def get_current_counts
...
 
+    def update_ralf_hostname environment, cloud
+      ralfs = find_nodes(roles: "clearwater-infrastructure", role: "ralf").length
+
+      changed_nodes = []
+
+      %w{bono ibcf sprout ralf}.each do |node_type|
+        find_nodes(roles: "clearwater-infrastructure", role: node_type).each do |node|
+          has_ralf = node.set[:clearwater][:ralf]
+          puts "#{node.name}: ralf attribute is #{has_ralf} and number of ralfs is #{ralfs}"
```
Comment (bossmc) [inline plugins/knife/knife-deployment-resize.rb]: Should be `Chef::Log` (throughout)

Comment (rkd-msw) [inline plugins/knife/knife-deployment-resize.rb]: quack quack

File: plugins/knife/knife-deployment-resize.rb (hunk 3 above)
Comment (rkd-msw) [inline plugins/knife/knife-deployment-resize.rb]: `default` is a bad choice for mutable attributes, per https://github.com/pikesley/learn-yourself-some-chef/wiki/Understanding-attributes:

> So it seems that what started out as a 'default' attribute from the cookbook has been nailed in as a 'normal' value on the node, and as a 'normal' attribute has higher precedence than 'default', further changes to the 'default' cookbook attribute will be silently ignored. I presume there are good reasons for this behaviour, but whatever, my advice for now is stay away from default attributes.

(I tried setting it to "default" and saw exactly these problems.)

**Pull request merged**